runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java

```diff
@@ -338,29 +342,55 @@ State waitUntilFinish(
     return null;  // Timed out.
   }
 
+  private AtomicReference<FutureTask<State>> cancelState = new AtomicReference<>();
+
   @Override
   public State cancel() throws IOException {
-    Job content = new Job();
-    content.setProjectId(getProjectId());
-    content.setId(jobId);
-    content.setRequestedState("JOB_STATE_CANCELLED");
-    try {
-      dataflowClient.updateJob(jobId, content);
-      return State.CANCELLED;
-    } catch (IOException e) {
-      State state = getState();
-      if (state.isTerminal()) {
-        LOG.warn("Job is already terminated. State is {}", state);
-        return state;
-      } else {
-        String errorMsg = String.format(
-            "Failed to cancel job in state %s, "
-                + "please go to the Developers Console to cancel it manually: %s",
-            state,
-            MonitoringUtil.getJobMonitoringPageURL(getProjectId(), getJobId()));
-        LOG.warn(errorMsg);
-        throw new IOException(errorMsg, e);
+    // Enforce that a cancel() call on the job is done at most once - as
+    // a workaround for Dataflow service's current bugs with multiple
+    // cancellation, where it may sometimes return an error when cancelling
+    // a job that was already cancelled, but still report the job state as
+    // RUNNING.
+    // To partially work around these issues, we absorb duplicate cancel()
+    // calls. This, of course, doesn't address the case when the job terminates
+    // externally almost concurrently to calling cancel(), but at least it
+    // makes it possible to safely call cancel() multiple times and from
+    // multiple threads in one program.
+    FutureTask<State> tentativeCancelTask = new FutureTask<>(new Callable<State>() {
+      @Override
+      public State call() throws Exception {
+        Job content = new Job();
+        content.setProjectId(getProjectId());
+        content.setId(jobId);
+        content.setRequestedState("JOB_STATE_CANCELLED");
+        try {
+          dataflowClient.updateJob(jobId, content);
+          return State.CANCELLED;
+        } catch (IOException e) {
+          State state = getState();
+          if (state.isTerminal()) {
+            LOG.warn("Job is already terminated. State is {}", state);
+            return state;
+          } else {
+            String errorMsg = String.format(
+                "Failed to cancel job in state %s, "
+                    + "please go to the Developers Console to cancel it manually: %s",
+                state,
+                MonitoringUtil.getJobMonitoringPageURL(getProjectId(), getJobId()));
+            LOG.warn(errorMsg);
+            throw new IOException(errorMsg, e);
+          }
+        }
       }
+    });
+    if (cancelState.compareAndSet(null, tentativeCancelTask)) {
+      // This is the thread that requested cancel first.
```

Comment:
I don't like the phrasing of "requested cancel first"

Maybe just something about getting the lock. This thread "won", regardless of request order.